core/google/api/core/gapic_v1/method.py

```diff
@@ -0,0 +1,193 @@
+    arguments.
+
+    For example::
+
+        import google.api.core.gapic_v1.method
+        from google.api.core import retry
+        from google.api.core import timeout
+
+        # The original RPC method.
+        def get_topic(name, timeout=None):
+            request = publisher_v2.GetTopicRequest(name=name)
+            return publisher_stub.GetTopic(request, timeout=timeout)
+
+        default_retry = retry.Retry(deadline=60)
+        default_timeout = timeout.Timeout(deadline=60)
+        wrapped_get_topic = google.api.core.gapic_v1.method.wrap_method(
+            get_topic, default_retry)
+
+        # Execute get_topic with default retry and timeout:
+        response = wrapped_get_topic()
+
+        # Execute get_topic without doing any retying but with the default
+        # timeout:
+        response = wrapped_get_topic(retry=False)
+
+        # Execute get_topic but only retry on 5xx errors:
+        my_retry = retry.Retry(retry.if_exception_type(
+            exceptions.InternalServerError))
+        response = wrapped_get_topic(retry=my_retry)
+
+    The way this works is by late-wrapping the given function with the retry
+    and timeout decorators. Essentially, when ``wrapped_get_topic()`` is
+    called:
+
+    * ``get_topic()`` is first wrapped with the ``timeout`` into
+      ``get_topic_with_timeout``.
+    * ``get_topic_with_timeout`` is wrapped with the ``retry`` into
+      ``get_topic_with_timeout_and_retry()``.
+    * The final ``get_topic_with_timeout_and_retry`` is called passing through
+      the ``args``  and ``kwargs``.
+
+    The callstack is therefore::
+
+        method.__call__() ->
+            Retry.__call__() ->
+                Timeout.__call__() ->
+                    wrap_errors() ->
+                        get_topic()
+
+    Note that if ``timeout`` or ``retry`` is ``False``, then they are not
+    applied to the function. For example,
+    ``wrapped_get_topic(timeout=False, retry=False)`` is more or less
+    equivalent to just calling ``get_topic`` but with error re-mapping.
+
+    Args:
+        func (Callable[Any]): The function to wrap. If timeout is specified,
+            then this function should take an optional timeout parameter.
+        default_retry (Optional[google.api.core.Retry]): The default retry
+            strategy.
+        default_timeout (Optional[google.api.core.Timeout]): The default
+            timeout strategy. If an int or float is specified, the timeout will
+            always be set to that value.
+        metadata (Union(Mapping[str, str], False)): A dict of metadata keys and
+            values. This will be augmented with common ``x-google-api-client``
+            metadata. If False, metadata will not be passed to the function
+            at all.
+
+    Returns:
+        TODO
+    """
+    func = grpc_helpers.wrap_errors(func)
+    if metadata is not False:
+        metadata = _prepare_metadata(metadata)
+
+    def method(*args, **kwargs):
```

Comment:
This closure has **a lot** going on. I'd be much more comfortable with a `__call__`-able object that owns the wrapped values `func`, `default_retry`, `default_timeout`, and `metadata`.

Reply:
Done.